### PR TITLE
Imlementing SQLDatabase.

### PR DESCRIPTION
**Pull request by nikolatzotchev**
Moved the methods to SQLDatabase, added the callbacks, probably not right but ilia can check. I don't now what kind of test I should write about this. The methods are already tested in the mock_database so no point in testing them, but maybe the callbacks?

### CODE REVIEW

**Comment by ammen99** [comment]
I think tests for the callbacks need to be implemented. As well as checking that they are called when the database is updated, and that no infinite recursion happens.

**Comment by nikolatzotchev** [comment]
> I think tests for the callbacks need to be implemented. As well as checking that they are called when the database is updated, and that no infinite recursion happens.

I'm not really sure how to test if there is no recursion happening.

**Comment by ammen99** [comment]
> I'm not really sure how to test if there is no recursion happening.

You can do the following:

1. Set a callback which increments a counter each time it is called
2. Update the database to callback is triggered
3. In the callback, increment your counter and update the database again
4. If the callback is called and the counter is higher than 0, this means you're in a recursion

**Comment by nikolatzotchev** [comment]
@ammen99 I fixed the bug here, it is not a problem right?

**Comment by nikolatzotchev** [comment]
#101 btw

**Comment by ammen99** [comment]
It's fine I think.

File: src/ja/server/database/sql/mock_database.py [hunk 1]
```diff
@@ -24,93 +19,15 @@ def __init__(self) -> None:
...
-        if since is None:
-            return jobs
-        return [job for job in jobs if job.statistics.time_added >= since]
-
-    def set_scheduler_callback(self, callback: ServerDatabase.RescheduleCallback) -> None:
-        pass
-
-    def set_job_status_callback(self, callback: ServerDatabase.JobStatusCallback) -> None:
-        pass
+    def __del__(self) -> None:
```
Comment (ammen99) [inline src/ja/server/database/sql/mock_database.py]: Is this needed? I thought Python will automatically call the parent's destructor.

File: src/ja/server/database/sql/mock_database.py [hunk 2]
```diff
@@ -24,93 +19,15 @@ def __init__(self) -> None:
...
-    def get_jobs_on_machine(self, machine: WorkMachine) -> Optional[List[Job]]:
-        jobs: Optional[List[Job]] = self.session.query(Job).join(DatabaseJobEntry). \
-            join(WorkMachine, WorkMachine.uid == machine.uid).all()
-        return jobs
-
-    def assign_job_machine(self, job: Job, machine: WorkMachine) -> None:
-        job_entry = self.find_job_by_id(job.uid)
-        job_entry.assigned_machine = machine
-        self.session.commit()
+        def stat_call(j: Job) -> None:
```
Comment (ammen99) [inline src/ja/server/database/sql/mock_database.py]: I think those things belong to the tests, not here.

File: src/test/server/database/test_database.py [hunk 3]
```diff
@@ -49,38 +52,17 @@ def setUp(self) -> None:
                                         WorkMachineResources(ResourceAllocation(12, 32, 12)))
         self.work_machine2 = WorkMachine("machi2", WorkMachineState.ONLINE,
                                          WorkMachineResources(ResourceAllocation(2, 2, 2)))
+        self.work_machine3 = WorkMachine("machi1", WorkMachineState.ONLINE,
```
Comment (ammen99) [inline src/test/server/database/test_database.py]: Is the UID intentionally the same as machine1 ?

File: src/test/server/database/test_database.py [hunk 4]
```diff
@@ -208,3 +196,40 @@ def test_query_jobs_false(self) -> None:
 
         jobs = self.mockDatabase.query_jobs(None, -1, self.work_machine)
         self.assertEqual(jobs, [self.mockDatabase.find_job_by_id(self.job.uid)])
+
+    def test_scheduler_callback(self) -> None:
+        call: Mock = Mock()
+        self.mockDatabase.set_scheduler_callback(call)
+        self.mockDatabase.update_job(self.job)
+        call.assert_called_with(self.mockDatabase)
```
Comment (ammen99) [inline src/test/server/database/test_database.py]: I think `assert_called_once_with` is even better.

Comment (ammen99) [reply src/test/server/database/test_database.py]: And in the other places as well.

File: src/ja/server/database/sql/database.py [hunk 5]
```diff
@@ -149,30 +151,121 @@ def __init__(self, host: str = None, user: str = None, password: str = None):
             "statistics": synonym("_statistics", descriptor=DatabaseJobEntry.statistics),
             "machine": synonym("_machine", descriptor=DatabaseJobEntry.assigned_machine)
         })
+        if user is not None and password is not None and host is not None:
+            # example: "postgresql://pesho:pesho@127.0.0.1:5432/jobadd"
+            conn: str = "postgresql://" + user + ":" + password + "@" + host + "/jobadder"
+            self.engine = create_engine(conn)
```
Comment (ammen99) [inline src/ja/server/database/sql/database.py]: private attributes should begin with `_`

File: src/test/server/database/test_database.py (hunk 3 above)
Comment (nikolatzotchev) [reply src/test/server/database/test_database.py]: yes, I just wanted to test when I update wm1 if it is going to break

File: src/ja/server/database/sql/mock_database.py (hunk 1 above)
Comment (nikolatzotchev) [reply src/ja/server/database/sql/mock_database.py]: I have no idea, but if that what are you telling me is true then it is not really needed.

**Pull request merged**